sound/soc/fsl/fsl_dai.c

```diff
@@ -0,0 +1,276 @@
+	if (priv->num_domains < 0) {
+		dev_err(&pdev->dev, "no power-domains property in %pOF\n", np);
+		return priv->num_domains;
+	}
+
+	priv->pd_dev = devm_kmalloc_array(&pdev->dev, priv->num_domains,
+					  sizeof(*priv->pd_dev), GFP_KERNEL);
+	if (!priv->pd_dev)
+		return -ENOMEM;
+
+	priv->link = devm_kmalloc_array(&pdev->dev, priv->num_domains,
+					sizeof(*priv->link), GFP_KERNEL);
+	if (!priv->link)
+		return -ENOMEM;
+
+	for (i = 0; i < priv->num_domains; i++) {
+		priv->pd_dev[i] = dev_pm_domain_attach_by_id(&pdev->dev, i);
+		if (IS_ERR(priv->pd_dev[i])) {
+			ret = PTR_ERR(priv->pd_dev[i]);
+			goto unroll_pm;
+		}
+
+		priv->link[i] = device_link_add(&pdev->dev, priv->pd_dev[i],
+						DL_FLAG_STATELESS |
+						DL_FLAG_PM_RUNTIME |
+						DL_FLAG_RPM_ACTIVE);
+		if (!priv->link[i]) {
+			ret = -EINVAL;
+			dev_pm_domain_detach(priv->pd_dev[i], false);
+			goto unroll_pm;
+		}
+	}
+
+	pm_runtime_enable(&pdev->dev);
+
+	ret = devm_snd_soc_register_component(&pdev->dev, &fsl_dai_component,
+					      priv->dai_drv, priv->num_drv);
+	if (ret < 0) {
+		dev_err(&pdev->dev, "Failed to register DAI ret = %d\n", ret);
+		return ret;
+	}
+	return 0;
+
+unroll_pm:
+	while (--i >= 0) {
+		device_link_del(priv->link[i]);
+		dev_pm_domain_detach(priv->pd_dev[i], false);
+	}
+	return ret;
+}
+
+static int fsl_dai_remove(struct platform_device *pdev)
+{
+	struct fsl_dai *priv = platform_get_drvdata(pdev);
+	int i;
+
+	pm_runtime_disable(&priv->pdev->dev);
+
+	for (i = 0; i < priv->num_domains; i++) {
+		device_link_del(priv->link[i]);
+		dev_pm_domain_detach(priv->pd_dev[i], false);
+	}
+
+	return 0;
+}
+
+static const struct of_device_id fsl_dai_dt_ids[] = {
+	{ .compatible = "fsl,esai-dai", },
+	{ .compatible = "fsl,sai-dai", },
+	{}
+};
+MODULE_DEVICE_TABLE(of, fsl_dai_dt_ids);
+
+#ifdef CONFIG_PM
+static int fsl_dai_runtime_resume(struct device *dev)
+{
+	struct fsl_dai *priv = dev_get_drvdata(dev);
+	int i, ret;
+
+	for (i = 0; i < priv->num_clks; i++) {
+		ret = clk_prepare_enable(priv->clks[i]);
+		if (ret < 0) {
+			dev_err(dev, "Failed to enable clk %s\n",
+				priv->clk_names[i]);
+			goto out;
+		}
+	}
+	return 0;
+out:
+	while (--i >= 0)
+		clk_disable_unprepare(priv->clks[i]);
+
+	return ret;
+}
+
+static int fsl_dai_runtime_suspend(struct device *dev)
+{
+	struct fsl_dai *priv = dev_get_drvdata(dev);
+	int i;
+
+	for (i = 0; i < priv->num_clks; i++)
+		clk_disable_unprepare(priv->clks[i]);
+
+	return 0;
+}
+#endif /* CONFIG_PM */
+
+static const struct dev_pm_ops fsl_dai_pm_ops = {
+	SET_RUNTIME_PM_OPS(fsl_dai_runtime_suspend,
+			   fsl_dai_runtime_resume, NULL)
+	SET_SYSTEM_SLEEP_PM_OPS(pm_runtime_force_suspend,
+				pm_runtime_force_resume)
+};
+
+static struct platform_driver fsl_dai_driver = {
+	.probe = fsl_dai_probe,
+	.remove = fsl_dai_remove,
+	.driver = {
+		.name = "fsl-dai",
+		.pm = &fsl_dai_pm_ops,
+		.of_match_table = fsl_dai_dt_ids,
+	},
+};
+
+module_platform_driver(fsl_dai_driver);
```

Comment:
what is the benefit of creating a platform device? couldn't you just use these functions as a library called from the i.MX probe functions?

Also if you have a child device which controls dependencies that the parent relies on (clock/power), this could be interesting from a suspend/resume perspective, where you can only suspend the parent once the child is suspended?

Reply:
@plbossart  In fact this is was how I initially started. Resources were managed from the i.MX probe functions.

Then I thought that a more natural way is to emulate the real DAIs in ALSA. This works nicely with simple-audio card machine driver.

Here is how the DT description of machine driver will look like:

```
       sof-audio-cs42888 {
»       »       compatible = "simple-audio-card";
»       »       simple-audio-card,dai-link {
»       »       »       format = "i2s";
»       »       »       cpu {
»       »       »       »       sound-dai = <&esai0_port>;
»       »       »       };
»       »       »       plat {
»       »       »       »       sound-dai = <&dsp>;
»       »       »       };
»       »       »       codec {
»       »       »       »       sound-dai = <&cs42888>;
»       »       »       };
»       »       };
»       };
```

Here's a sample of the DAI DT description:

```
»       esai0_port: esai-port {
»       »       #sound-dai-cells = <0>;
»       »       compatible = "fsl,esai-dai";
»       »       pinctrl-names = "default";
»       »       pinctrl-0 = <&pinctrl_esai0>;
»       »       power-domains = <&pd IMX_SC_R_ESAI_0>, <&pd IMX_SC_R_DMA_0_CH6>..;
»       »       clocks = <&esai0_lpcg 1>, <&esai0_lpcg 0>,  <&esai0_lpcg 1>, <&clk_dummy>;
»       »       clock-names = "core", "extal", "fsys", "spba";
»       »       status="okay";
»       };

Reply:
I guess I am not clear on the relationship between your SOF device and your ESAI device? If they are completely independent this could be tricky for power management.

Reply:
SOF device will act as a platform DAI device in ALSA, and ESAI will act as a CPU DAI in ALSA.

Here is how to sound card will look like:

```
wm8960-sound {
      cpu-dai {
            sound-dai = <esai>
      };
     platform-dai {
           sound-dai = <dsp_device>;
     };
     codec = {
          sound-dai = <wm8960>;
}
```

I must admit that I'm yet to have a deeper look at PM, but I'm basing my proposal on the fact that this scenario with dai, platform, codec already works in ALSA.

Reply:
> @plbossart
> 
> SOF device will act as a platform DAI device in ALSA, and ESAI will act as a CPU DAI in ALSA.
> 
> Here is how to sound card will look like:
> 
> ```
> wm8960-sound {
>       cpu-dai {
>             sound-dai = <esai>
>       };
>      platform-dai {
>            sound-dai = <dsp_device>;
>      };
>      codec = {
>           sound-dai = <wm8960>;
> }
> ```
> 
> I must admit that I'm yet to have a deeper look at PM, but I'm basing my proposal on the fact that this scenario with dai, platform, codec already works in ALSA.
> 
> So, SOF device and ESAI device are not completely independent. They are siblings to the machine device node.

I am officially confused, I barely understood what a platform component was (traditionally for DMA, not sure why it's needed for Intel and NXP). I have never heard of a platform-dai,

Reply:
platform dai is the same thing as platform component. I named it platform dai because is encountered near cpu-dai / codec-dai